crates/rrg/src/action/get_tcp_response.rs

```diff
@@ -0,0 +1,335 @@
+
+/// List of possible failure scenarios for `get_tcp_response`.
+#[derive(Debug)]
+enum ErrorKind {
+    Connect,
+    Write,
+    Read,
+    ShutdownWrite,
+    ShutdownRead,
+    InvalidWriteTimeout,
+    InvalidReadTimeout,
+}
+
+/// Error type for failures specific to `get_tcp_response`.
+#[derive(Debug)]
+struct Error {
+    kind: ErrorKind,
+    inner: std::io::Error,
+}
+
+impl std::fmt::Display for ErrorKind {
+
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Connect => write!(fmt, "unable to connect"),
+            Self::Write => write!(fmt, "unable to write to stream"),
+            Self::Read => write!(fmt, "unable to read from stream"),
+            Self::ShutdownWrite => write!(fmt, "failed to shutdown writing"),
+            Self::ShutdownRead => write!(fmt, "failed to shutdown reading"),
+            Self::InvalidWriteTimeout => write!(fmt, "invalid write timeout"),
+            Self::InvalidReadTimeout => write!(fmt, "invalid read timeout"),
+        }
+    }
+}
+
+impl std::fmt::Display for Error {
+
+    fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(fmt, "{}: {}", self.kind, self.inner)
+    }
+}
+
+impl std::error::Error for Error {
+
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        Some(&self.inner)
+    }
+}
+
+impl From<Error> for crate::session::Error {
+
+    fn from(error: Error) -> crate::session::Error {
+        crate::session::Error::action(error)
+    }
+}
+
+impl crate::request::Args for Args {
+
+    type Proto = rrg_proto::get_tcp_response::Args;
+
+    fn from_proto(mut proto: Self::Proto) -> Result<Self, crate::request::ParseArgsError> {
+        use crate::request::ParseArgsError;
+
+        let addr = std::net::SocketAddr::try_from(proto.take_address())
+            .map_err(|error| {
+                ParseArgsError::invalid_field("address", error)
+            })?;
+
+        let connect_timeout = std::time::Duration::try_from(proto.take_connect_timeout())
+            .map_err(|error| {
+                ParseArgsError::invalid_field("connect timeout", error)
+            })?;
+        let write_timeout = std::time::Duration::try_from(proto.take_write_timeout())
+            .map_err(|error| {
+                ParseArgsError::invalid_field("write timeout", error)
+            })?;
+        let read_timeout = std::time::Duration::try_from(proto.take_read_timeout())
+            .map_err(|error| {
+                ParseArgsError::invalid_field("read timeout", error)
+            })?;
+
+        Ok(Args {
+            addr,
+            connect_timeout,
+            write_timeout,
+            read_timeout,
+            data: proto.take_data(),
+        })
+    }
+}
+
+impl crate::response::Item for Item {
+
+    type Proto = rrg_proto::get_tcp_response::Result;
+
+    fn into_proto(self) -> Self::Proto {
+        let mut proto = rrg_proto::get_tcp_response::Result::new();
```

Comment:
Would it make sense to add a field to the result proto that indicates if the response was larger than 1MiB and cut off?

Reply:
Yeah, good question. On one hand it makes sense, on the other hand if we decide to send multiple responses in the future it feels a bit weird.

I added a TODO comment to consider this in the future but for now I will skip it not to complicate the implementation further since we have no need for this case at the moment anyway (we only need this action for cloud VM metadata collection which should be small HTTP response).